docs/explanations/implementation.md

```diff
@@ -0,0 +1,78 @@
+- `morphoTip`: Points to the latest OBFT block that was applied to the state, doesn't include transactions in the mempool
+- `lastCheckpoint`: Contains the last checkpoint created, which stores the proof-of-work block that was voted for and the signatures of a majority of nodes
+- `checkpointAt`: Points to the OBFT block that created the checkpoint in `lastCheckpoint`. This allows distinguishing between older checkpoints and ones just created, used to determine whether it should be [pushed](#pushing-new-checkpoints) or not
+- `currentVotes`: For each federation node, this stores either
+  - No vote if they haven't cast a vote since the last checkpoint was created
+  - The latest vote they cast since the checkpoint was created
+
+### Checkpoint
+
+[(source)](../../morpho-checkpoint-node/src/Morpho/Ledger/PowTypes.hs#L48-L53)
+
+A checkpoint as stored in the `lastCheckpoint` field of the [ledger state](#ledger-state-data-structures) contains the following fields:
+- `checkpointedBlock`: The proof-of-work block that was checkpointed
+- `chkpSignatures`: A list of signatures from a majority of nodes (counting [`RequiredMajority`](../references/configuration.md#requiredmajority)). See the [crypto document](../explanations/crypto.md#vote-signing) for the components of these signatures.
+
+### Vote
+
+[(source)](../../morpho-checkpoint-node/src/Morpho/Ledger/PowTypes.hs#L41-L46)
+
+A vote as stored in `currentVotes` field of the [ledger state](#ledger-state-data-structures) contains the following fields:
+- `votedPowBlock`: The proof-of-work block that a node voted for as the next block to be checkpointed
+- `voteSignature`: A single signature by the node that cast the vote. See the [crypto document](../explanations/crypto.md#vote-signing) for the components of such signatures.
+
+## Ledger state transitions
+
+[(source)](../../morpho-checkpoint-node/src/Morpho/Ledger/Update.hs#L251-L281)
+
+In OBFT, time is divided into slots, whose duration is controlled with [`SlotDuration`](../references/configuration.md#slotduration). At every slot, the node whose [`NodeId`](../references/configuration.md#nodeid) equals the slot number modulo [`NumCoreNodes`](../references/configuration.md#numcorenodes) is the leader. This node collects transactions from all the other nodes into a mempool, then forges an OBFT block from them, which is then published to all the other nodes. In Morpho's case, a transaction is equivalent to a [vote](#vote), and an OBFT block consists of an arbitrary number of transactions/votes.
+
+As new OBFT blocks are forged, the [ledger state](#ledger-state-data-structures) changes as follows:
+- `morphoTip` is updated to the new OBFT block's tip. This ensures that when [Pushing a checkpoint to Mantis](#mantis-pushing), the same checkpoint isn't pushed multiple times [2]
+- All the blocks transactions/votes are inserted into `currentVotes`, overriding previous votes by the same public key [3]
```

Comment:
There's three different things going on:
- There is the _mempool_, which the slot leader uses to collect all transactions/votes from the other nodes. This acts as a sort of staging area for votes, where they are just a potential candidate to be included in the next block.
- The slot leader then forges the block from the transactions in the mempool. Over time this gives you a _chain of blocks_, where each block is just a list of votes in our case.
- And finally, if we think of a chain of blocks as a sequence of state updates, where each block contains information on how the state should change. With this we can calculate a _ledger state_ for a specific chain of blocks.

Here `currentVotes` is part of the ledger state, containing most recent votes from blocks that haven't been included in the last checkpoint. While blocks themselves only contain votes, we can calculate a ledger state for each of them (over the block and its predecessors). This can then be thought of as a chain of ledger states next to a chain of blocks. This looks something like this (with `RequiredMajority = 3`):

| Block | State.`currentVotes` | State.`lastCheckpoint` |
| --- | --- | --- |
| Genesis | [] | \<none\> |
| Block1(v1) | [v1] | \<none\> |
| Block2(v2,v3) | [] | Checkpoint(\<block\>, votes: [v1, v2, v3]) |
| Block3(v4, v5) | [v4, v5] | Checkpoint(\<block\>, votes: [v1, v2, v3]) |
```

I'll probably make a diagram for this

Reply:
Thanks Silvan! Very clear now. I would be awesome if this was included in the docs

Reply:
I might still do this, not yet sure how to go about visualizing this nicely